### ISSUE TITLE
Mandatory logging procedures use GNU-only shell syntax, fail on macOS

### ISSUE THREAD

**Issue by monroyd** [rebelytics/one-skill-to-rule-them-all#16]
The "How to Log" section labels the write-time assertion and post-write verification blocks **mandatory**, but both use GNU-only syntax:

- `grep -oP '### Observation \K\d+'` — macOS/BSD grep has no `-P` (fails with `invalid option -- P`)
- `sed -i "..."` — BSD sed requires an explicit backup suffix (`sed -i ''`), so the bare form errors out

The pre-logging one-liner does offer a POSIX alternative, but the two mandatory blocks have none — so on macOS the mandated procedure cannot be executed as written.

**Suggested fix:** make all snippets portable, e.g.

```bash
command grep -o '### Observation [0-9]*' log.md | command grep -o '[0-9]*' | sort -n | tail -1
```

and note the `sed -i ''` (BSD) vs `sed -i` (GNU) difference. (`command grep` also sidesteps shell-function shadowing of `grep`, which occurs in some agent harnesses.) Verified working end-to-end on macOS, including the collision-renumber path.

---

**Optional, separate from the bug above:** the skill is ~9.5k words and is designed to load at the start of every task session, so every non-operational line is paid on each session. By the skill's own Lean Content rule, the licensing survey, attribution template, handoff-doc format, and review rationale could live in a companion reference doc loaded on demand, cutting per-session cost roughly in half or more.

Filed from an install's skill-observation log (Observations 1 and a portability note), not a live bug report against a specific run.

**Comment by rebelytics** [comment]
Thanks for this, @monroyd. As you’ve provided a suggested fix, I will create a PR with you as a co-author for the next version of task-observer. No action required on your side. And thanks for the additional input on further improving progressive disclosure. This is also panned for v3.

**Comment by rebelytics** [comment]
Sorry, @monroyd and @jazorro11, for getting confused in my previous comment (edited now). This is a clear sign that I should call it a day, rest, and continue next week. Thanks for your contributions and have a nice weekend!

**Comment by rebelytics** [comment]
Fixed in `release/v3.0.0`, mostly by removal: the GNU-only numbering and renumbering blocks no longer exist, because 3.0.0 stores one file per observation and the numbering ritual went with the shared log. Every remaining snippet is POSIX. Your lean-content aside is addressed too — the core is reorganised and the rationale moved into reference files.

You're credited as co-author on the commit. FYI: I'll be running v3 for a week or so before it goes into main.

**Comment by certen59** [comment]
Adding a third failure mode to this, from a Windows install, plus a suggestion
that generalises past patching the snippets one platform at a time.

**The PowerShell case is a different failure class from the two already
reported here.** #16 is "the snippet errors out" (BSD grep rejects `-P`); #25 is
the nastier "the snippet succeeds and returns nothing" (Git-Bash locale). On a
host whose primary shell is PowerShell, neither happens, because the snippet
never runs at all: 5.1 has no `grep`, and `&&` is a *parse* error rather than a
runtime one, so the whole line is rejected before anything executes. There is no
stdout to misread and no error to interpret — there is just a procedure the
agent cannot follow.

What happens next is the actual problem. Faced with a mandatory step it cannot
execute, an agent either improvises a half-version or quietly drops the checks
and appends by remembered number. Neither is visible at the time. The collision
surfaces sessions later in a shared log — exactly the failure the discipline
exists to prevent, arrived at by way of the discipline being unusable.

**Suggestion, in addition to making the snippets portable:** separate the rule
from its example implementation, so the rule survives an environment where no
provided snippet runs. Two parts.

In `SKILL.md`, under `**Numbering discipline (mandatory, every append):**`:

> The three steps are the rule; the snippets are one POSIX example of it. In a
> shell that lacks them — a PowerShell-only Windows install, a restricted
> runner — do not skip the discipline and do not improvise a half-version of
> it. Run the same check-act-verify sequence through whatever scripting the
> environment does have; `references/environments.md` → Shell portability
> carries a shell-agnostic implementation. A skipped numbering check fails
> silently: nothing goes wrong until a collision surfaces sessions later.

And a `## Shell portability` section in `references/environments.md` carrying the
episodic detail plus a stdlib-only Python equivalent. Run against a copy of a
real log — happy path, collision branch, and the zero branch:

```python
import re, pathlib
HEADER = re.compile(r'(?m)^### Observation (\d+):')
log = pathlib.Path("log.md")

# 1. Pre-check — read the live file, never session memory
existing = [int(n) for n in HEADER.findall(log.read_text(encoding="utf-8"))]
proposed = max(existing) + 1 if existing else 1

# 2. Collision assertion — immediately before the append
if proposed in existing:
    raise SystemExit(f"COLLISION on #{proposed}")

# 3. Append — build the entry FROM `proposed` so the number cannot drift
entry = f"\n### Observation {proposed}: {title}\n\n**Status:** OPEN\n{body}"
before = len(existing)
with log.open("a", encoding="utf-8") as f:
    f.write(entry)

# 4. Post-write verification against the live file
text = log.read_text(encoding="utf-8")
assert len(HEADER.findall(text)) == before + 1, "header count off"
mine = len(re.findall(rf'(?m)^### Observation {proposed}:', text))
assert mine != 0, f"#{proposed} never landed — the entry carried a stale number"
assert mine == 1, f"collision: #{proposed} appears {mine}x — renumber yours to max+1"
```

Python is not the point — any language that can read the file works. The point
is that the header-count invariant in Log-write safety is a property of the
file, not of `grep`, and saying so in the docs stops the next
environment-specific report from being needed.

(The `mine != 0` assertion covers a separate defect in step 3 as written —
filed as #67 so this thread stays about portability.)

Principle: stating a mandatory discipline in one shell's syntax makes the
discipline depend on that shell. The thing that gets dropped when the
environment changes is then the rule, not the example.

---

🤖 Prepared with [Claude Code](https://claude.com/claude-code). Written by the
agent from a real install; the human operator reviewed and approved the
submission.

**Comment by rebelytics** [comment]
@certen59: thank you for this genuinely useful taxonomy: the snippet that errors (BSD grep), the snippet that succeeds emptily (locale), and now the snippet that never parses (PowerShell). Your PowerShell class and the generalise-past-per-platform-patching suggestion deserve their own treatment: I've opened a dedicated issue for the platform-portability pass (cross-referenced just above in this thread, together with #64, #66 and #72–74) rather than patching snippets one shell at a time.

Closing this issue with v3.0.0, released today: the GNU-only blocks it was about are gone (the per-file log removed the numbering ritual entirely) and the remaining snippets are POSIX.

@monroyd: thanks again for the report that started this thread; your suggested fix shaped where it ended up.
